habitat_baselines/rl/ppo/ppo_trainer.py

```diff
@@ -0,0 +1,519 @@
+import os
+import time
+from collections import deque
+from typing import Dict, List
+
+import numpy as np
+import torch
+
+from habitat import logger
+from habitat.utils.visualizations.utils import observations_to_image
+from habitat_baselines.common.base_trainer import BaseRLTrainer
+from habitat_baselines.common.baseline_registry import baseline_registry
+from habitat_baselines.common.env_utils import construct_envs
+from habitat_baselines.common.environments import NavRLEnv
+from habitat_baselines.common.tensorboard_utils import get_tensorboard_writer
+from habitat_baselines.common.utils import (
+    RolloutStorage,
+    batch_obs,
+    generate_video,
+    poll_checkpoint_folder,
+    update_linear_schedule,
+)
+from habitat_baselines.rl.ppo import PPO, Policy
+
+
+@baseline_registry.register_trainer(name="ppo")
+class PPOTrainer(BaseRLTrainer):
+    supported_tasks = ["Nav-v0"]
+
+    def __init__(self, config=None):
+        super().__init__(config)
+        print(config)
+        self.actor_critic = None
+        self.agent = None
+        self.envs = None
+        self.device = None
+        if config is not None:
+            logger.info(f"env config: {config}")
+            self.envs = construct_envs(config, NavRLEnv)
+            self._setup()
+
+    def _setup(self):
+        ppo_cfg = self.config.TRAINER.RL.PPO
+        self.device = torch.device("cuda", ppo_cfg.pth_gpu_id)
+        logger.add_filehandler(ppo_cfg.log_file)
+        if not os.path.isdir(ppo_cfg.checkpoint_folder):
+            os.makedirs(ppo_cfg.checkpoint_folder)
+
+        self.actor_critic = Policy(
+            observation_space=self.envs.observation_spaces[0],
+            action_space=self.envs.action_spaces[0],
+            hidden_size=512,
+            goal_sensor_uuid=self.config.TASK.GOAL_SENSOR_UUID,
+        )
+        self.actor_critic.to(self.device)
+
+        self.agent = PPO(
+            self.actor_critic,
+            ppo_cfg.clip_param,
+            ppo_cfg.ppo_epoch,
+            ppo_cfg.num_mini_batch,
+            ppo_cfg.value_loss_coef,
+            ppo_cfg.entropy_coef,
+            lr=ppo_cfg.lr,
+            eps=ppo_cfg.eps,
+            max_grad_norm=ppo_cfg.max_grad_norm,
+        )
+
+    def save_checkpoint(self, file_name: str):
+        checkpoint = {
+            "state_dict": self.agent.state_dict(),
+            "config_str": str(self.config),
+        }
+        torch.save(
+            checkpoint,
+            os.path.join(
+                self.config.TRAINER.RL.PPO.checkpoint_folder, file_name
+            ),
+        )
+
+    def load_checkpoint(self, checkpoint_path, *args, **kwargs) -> Dict:
+        return torch.load(checkpoint_path, map_location=self.device)
+
+    def train(self):
+        assert (
```

Comment:
The train method for PPO is really heavy, we should abstract out logic present inside this class to separate methods.